### PR TITLE
Callback errors

### PR DESCRIPTION
**Pull request by identitysolutions**
We have a customer project that requires Drupal 7 (PR #124) & Drupal 8 SSO with Auth0.
This PR is a result of the changes required to function with the latest Auth0 APIs, e.g Management API changes.

A summary of this PR is:

- ~Updated verify_email to use v2 Management API.~
- Fix an issue with callback when a consent was required, handle other errors like UnauthorizedError thrown from a Rule.
- Add an option for Custom Domain config setting.

### CODE REVIEW

File: src/Controller/AuthController.php [hunk 1]
```diff
@@ -87,6 +89,13 @@ class AuthController extends ControllerBase {
    */
   protected $domain;
 
+  /**
+   * The Auth0 Custom Domain.
+   *
+   * @var string|null
+   */
+  protected $custom_domain;
```
Comment (joshcanhelp) [inline src/Controller/AuthController.php]: Let's use snakeCase for properties in this class.

File: src/Controller/AuthController.php [hunk 2]
```diff
@@ -376,18 +386,32 @@ protected function buildAuthorizeUrl($prompt, $returnTo = NULL) {
...
+          $request->query->get('error') == 'consent_required') {
+        return new TrustedRedirectResponse($this->buildAuthorizeUrl(FALSE, $returnTo));
+      } else {
+        return $this->failLogin($error_msg . $request->query->get('error_description'), $request->query->get('error_description'));
+      }
     }
     // Check in post.
-    if ($request->request->has('error') && $request->request->get('error') == 'login_required') {
-      return new TrustedRedirectResponse($this->buildAuthorizeUrl(FALSE, $returnTo));
+    if ($request->request->has('error')) {
```
Comment (joshcanhelp) [inline src/Controller/AuthController.php]: Wouldn't `request` include POST and GET? 

Also ... I don't think this module uses anything besides an authorization code in the URL query.

File: src/Controller/AuthController.php [hunk 3]
```diff
@@ -1044,23 +1056,23 @@ public function verify_email(Request $request) {
 
     try {
       $userId = $user->sub;
-      $url = "https://$this->domain/api/users/$userId/send_verification_email";
-
-      $client = $this->httpClient;
+      /* Need to use v2 Mgmt API to send verify email, must use tenancy domain for audience */
+      $audience = 'https://' . $this->domain . '/api/v2/';
+      $auth0Api = new Authentication($this->helper->getDomain(), $this->clientId, $this->clientSecret, $audience);
+      $auth0Client = $auth0Api->client_credentials([]);
```
Comment (joshcanhelp) [inline src/Controller/AuthController.php]: Two problems here:

1. The Application might not have the Client Credentials grant allowed
2. The Application might not be authorized for the Management API (it's not by default)

Without any error-checking here, this process will fail and I'm not sure what that would look like for the user.

File: src/Controller/AuthController.php [hunk 4]
```diff
@@ -376,18 +386,32 @@ protected function buildAuthorizeUrl($prompt, $returnTo = NULL) {
...
    *   The return url.
    *
-   * @return \Drupal\Core\Routing\TrustedRedirectResponse|null
+   * @return \Drupal\Core\Routing\TrustedRedirectResponse
+   *         \Symfony\Component\HttpFoundation\RedirectResponse
+   *         null
    *   The redirect response.
    */
   private function checkForError(Request $request, $returnTo) {
+    $error_msg = $this->t('There was a problem logging you in, sorry for the inconvenience.');
```
Comment (joshcanhelp) [inline src/Controller/AuthController.php]: Let's drop "sorry for the inconvenience" here

File: src/Controller/AuthController.php [hunk 5]
```diff
@@ -376,18 +386,32 @@ protected function buildAuthorizeUrl($prompt, $returnTo = NULL) {
...
     // Check for errors.
     // Check in query.
-    if ($request->query->has('error') && $request->query->get('error') == 'login_required') {
-      return new TrustedRedirectResponse($this->buildAuthorizeUrl(FALSE, $returnTo));
+    if ($request->query->has('error')) {
+      if ($request->query->get('error') == 'login_required' || 
+          $request->query->get('error') == 'consent_required') {
+        return new TrustedRedirectResponse($this->buildAuthorizeUrl(FALSE, $returnTo));
+      } else {
+        return $this->failLogin($error_msg . $request->query->get('error_description'), $request->query->get('error_description'));
```
Comment (joshcanhelp) [inline src/Controller/AuthController.php]: Need a space in between the concatenated strings.

File: src/Form/BasicSettingsForm.php [hunk 6]
```diff
@@ -61,6 +61,14 @@ public function buildForm(array $form, FormStateInterface $form_state) {
       '#required' => TRUE,
     ];
 
+    $form['auth0_custom_domain'] = [
+      '#type' => 'textfield',
+      '#title' => $this->t('Custom Domain'),
+      '#default_value' => $config->get('auth0_custom_domain') ?: '',
+      '#description' => $this->t('Your Auth0 custom domain, if in use. You can see it in the auth0 dashboard.'),
```
Comment (joshcanhelp) [inline src/Form/BasicSettingsForm.php]: Let's remove "You can see it in the auth0 dashboard."

**Comment by joshcanhelp** [comment]
@identitysolutions - Are you able to address the conflict and comments here? We're working on a release now and would love to get your contribution in as part of that.

Thanks in advance!

File: src/Controller/AuthController.php (hunk 3 above)
Comment (identitysolutions) [reply src/Controller/AuthController.php]: This code is based on the Wordpress Auth0 module. It's wrapped in a try block and has UI info to the end user as well as logging to Drupal.

**Comment by identitysolutions** [comment]
> @identitysolutions - Are you able to address the conflict and comments here? We're working on a release now and would love to get your contribution in as part of that.
> 
> Thanks in advance!

I'll take a look. At lot of your comments are on existing code and not my PR

**Comment by identitysolutions** [comment]
Hi Josh,

I have update

**Comment by identitysolutions** [comment]
Hi Josh,

I have updated the branch, so this should now update the PR.

**Comment by joshcanhelp** [comment]
@identitysolutions - Almost there, thanks for your work here!

I think we should put the verify email changes in a different PR so it can be reviewed separately. The v1 API does need to be updated but it's working on my test install now and I don't want to break that out of the blue without consideration and a major version release. If you can pull that out, I'll test this locally and get it merged.

Thanks again!

**Comment by joshcanhelp** [comment]
Sorry, closed on accident!

**Comment by identitysolutions** [comment]
> @identitysolutions - Almost there, thanks for your work here!
> 
> I think we should put the verify email changes in a different PR so it can be reviewed separately. The v1 API does need to be updated but it's working on my test install now and I don't want to break that out of the blue without consideration and a major version release. If you can pull that out, I'll test this locally and get it merged.
> 
> Thanks again!

Hi Josh,
I have reverted the verify_email commit and will put in a separate PR.

**Comment by joshcanhelp** [comment]
@identitysolutions - Thanks again! I have this on my list to review and will get back to you before the end of the week.

File: src/Util/AuthHelper.php [hunk 7]
```diff
@@ -123,7 +126,21 @@ public function validateIdToken($idToken) {
...
+   *
+   * @param none
+   *
+   * @return mixed
+   *   A string with the domain name
+   *   A empty string if the config is not set
+   *
+   */
+  public function getDomain() {
+      return isset($this->customDomain) ? $this->customDomain : $this->domain;
```
Comment (joshcanhelp) [inline src/Util/AuthHelper.php]: This should be:

```suggestion
      return !empty($this->customDomain) ? $this->customDomain : $this->domain;
```

File: src/Controller/AuthController.php [hunk 8]
```diff
@@ -269,7 +278,7 @@ public function login(Request $request) {
         'drupalSettings' => [
           'auth0' => [
             'clientId' => $this->config->get('auth0_client_id'),
-            'domain' => $this->config->get('auth0_domain'),
+            'domain' => $this->helper->getDomain(),
             'lockExtraSettings' => $lockExtraSettings,
             'configurationBaseUrl' => 'https://cdn.auth0.com',
```
Comment (joshcanhelp) [inline src/Controller/AuthController.php]: This needs to take into account the tenant region. For `eu` and `au` regions, the base URL here needs to reflect that: 

```
// If tenant is US (no region in the URL):
'configurationBaseUrl' => 'https://cdn.auth0.com',

// If tenant is EU (tenant.eu.auth0.com):
'configurationBaseUrl' => 'https://cdn.eu.auth0.com',

// If tenant is AU (tenant.au.auth0.com):
'configurationBaseUrl' => 'https://cdn.au.auth0.com',
```

Here's what we do in the WordPress plugin:

https://github.com/auth0/wp-auth0/blob/master/WP_Auth0.php#L198

Comment (identitysolutions) [reply src/Controller/AuthController.php]: Are you sure this is necessary?

https://cdn.eu.auth0.com & https://cdn.au.auth0.com result in a 404 error from a HTTP GET.
https://cdn.us.auth0.com is missing a certificate.
The doco says you only need ```eu``` and ```au``` but the WP code returns us by default.

Comment (joshcanhelp) [reply src/Controller/AuthController.php]: > Are you sure this is necessary?

Yes:

https://auth0.com/docs/custom-domains/additional-configuration#embedded-lock

> https://cdn.eu.auth0.com & https://cdn.au.auth0.com result in a 404 error from a HTTP GET.

The base URL is never used. Those domains serve required configuration files that Lock uses. 

> https://cdn.us.auth0.com is missing a certificate.

That one is invalid, it's `cdn.auth0.com` for the US region.

> The doco says you only need eu and au but the WP code returns us by default.

There test for that method might be helpful to see  how that works:

https://github.com/auth0/wp-auth0/blob/master/tests/testWPAuth0Helpers.php#L18

Here's how the actual configurationBaseUrl is built:

https://github.com/auth0/wp-auth0/blob/master/tests/testLockOptions.php#L19

**Comment by identitysolutions** [comment]
Hi Josh,
Commit has been pushed.

Regards,

Gary Bromley

[cid:identity-solutions_12fd59f8-1ca3-40ae-ad5a-22e32bce7c74.png]
Identity Solutions Australia Pty Ltd
1408-1411 Level 14 Aquavista Tower
401 Docklands Drive Docklands VIC 3008
p.  +61 (3) 8420 3420  m.  +61 417 144 648
f.   +61 (3) 8420 3440  d.  +61 (3) 8420 3421
e.  gary.bromley@identity-solutions.com.au
w. https://www.identity-solutions.com.au

________________________________

**Pull request merged**